### ISSUE TITLE
New Lint for `vec![Rc::new(..); N]`

### ISSUE THREAD

**Issue by 0e4ef622** [rust-lang-nursery/rust-clippy#3377]
A guy on discord got pretty confused by this.

**Comment by flip1995** [comment]
Could you please give a little more information here? What should the Lint warn about or suggest? Maybe a code example would also be helpful.

**Comment by 0e4ef622** [comment]
The `vec!` macro uses Clone, and since Rc's clone doesn't duplicate the underlying data, all the elements of the Vec end up pointing to the same thing. This can cause unexpected issues with interior mutability.

```
use std::{rc::Rc, cell::Cell};
let v = vec![Rc::new(Cell::new(false)); 3];
v[1].set(true);
println!("{:?}", v); // [true, true, true]
```

**Comment by 0e4ef622** [comment]
I'm not sure about the _best_ alternative, `resize_with` seems like a good option but is currently unstable. Possibly something like,
```rust
(0..N).map(|_| Rc::new(..)).collect()
```
¯\\\_(ツ)\_/¯

**Comment by y21** [comment]
This lint exists today. Output for the code above:
```
warning: initializing a reference-counted pointer in `vec![elem; len]`
 --> src/main.rs:4:13
  |
4 |     let v = vec![Rc::new(Cell::new(false)); 3];
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: each element will point to the same `Rc` instance
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#rc_clone_in_vec_init
  = note: `#[warn(clippy::rc_clone_in_vec_init)]` on by default
```
